Ticket2539:  Switch to new NICOS protocol

DominicOram opened the pull request:
### Description of work

Modified the GUI to communicate with NICOS using ZeroMQ.

### To test

* Get the new NICOS core from \\olympic\Babylon5\Scratch\Dom Oram\nicos-core
* Run bin\nicos-daemon (I had some difficulty merging this into our nicos repo)
* Confirm that on the NICOS scripting screen this connects
* Confirm you can send a script on the NICOS screen
* Confirm that pointing to an instrument that isn't running NICOS gives a sensible message and periodically attempts to reconnect
* Confirm unit tests run

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [ ] Have no new checkstyle warnings been introduced? Check via Jenkins
- [ ] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [ ] Are there automated [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?
    - Manual system tests for the functionality should be added if there are no automated tests
    - Manual system tests can be removed from the template if they are covered by suitable automated tests
- [ ] Did any existing system test break as a result of the current changes? 
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)? There is a script called `check_opi_format.py` to help with this.

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has developer documentation been updated if required?
- [ ] Have any new plugins been added to the appropriate feature.xml? You can check if this is needed by validating plugins as per method at  https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Common-Eclipse-Issues

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

DominicOram commented:
Once this is reviewed we should:
- [ ] Update our NICOS repo to the latest version
- [ ] Remove the ISIS script server proxy
- [ ] Update the developers manual

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.nicos/src/uk/ac/stfc/isis/ibex/nicos/messages/NICOSMessage.java:
```diff
@@ -0,0 +1,61 @@
+     *             thrown when the conversion cannot take place.
+     */
+    public List<String> getMulti() throws ConversionException {
+        JsonSerialisingConverter<List<Object>> serialiser = new JsonSerialisingConverter<>(List.class);
```
Compile error: illegal conversion

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.nicos/src/uk/ac/stfc/isis/ibex/nicos/messages/Login.java:
```diff
@@ -21,14 +21,31 @@
+        Map<String, String> params = new HashMap<>();
+        params.put("login", "ibex");
+        params.put("passwd", "a2eed0a7fcb214a497052435191b5264cca5b687");
+        parameters = Arrays.asList(params);
```
Compile error: illegal conversion

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.nicos/src/uk/ac/stfc/isis/ibex/nicos/messages/QueueScript.java:
```diff
@@ -39,8 +40,14 @@
-        this.name = name;
-        this.code = code;
+        this.command = "queue";
+        this.parameters = Arrays.asList(name, code);
```
Compile error: illegal conversion

DominicOram commented:
Apologies, I forgot to switch my IDE back to java 7 after doing the ArchiveEngine work.

DominicOram commented on base/uk.ac.stfc.isis.ibex.nicos.tests/src/uk/ac/stfc/isis/ibex/nicos/ZMQSessionTest.java:
```diff
@@ -0,0 +1,231 @@
+    public void GIVEN_NICOS_message_with_five_parts_WHEN_message_sent_THEN_five_parts_sent() {
+        sendBlankMessage(5);
+
+        verify(zmq, times(5)).send(anyString(), anyBoolean());
```
Yh, you're right this is silly. I've kept the 3 parts with a comment as this is what most of the NICOS commands will be but the 5 parts isn't required

DominicOram commented on base/uk.ac.stfc.isis.ibex.nicos/src/uk/ac/stfc/isis/ibex/nicos/messages/QueueScript.java:
```diff
@@ -39,8 +40,14 @@
+    @Override
+    public ReceiveMessage parseResponse(String response) throws ConversionException {
+        JsonDeserialisingConverter<String> deserial = new JsonDeserialisingConverter<>(String.class);
+        return new ReceiveStringMessage(deserial.convert(response));
```
Sort of, the response that comes from the server will be a JSON string and so will have quotation marks in and potentially some whitespace, this will strip those.

DominicOram commented on base/uk.ac.stfc.isis.ibex.nicos.tests/src/uk/ac/stfc/isis/ibex/nicos/NicosModelTest.java:
```diff
@@ -0,0 +1,414 @@
+    }
+
+    @Test
+    public void GIVEN_failing_session_connection_WHEN_connect_first_called_THEN_first_status_is_connecting() {
```
Added a similar one for successful connections

DominicOram commented:
* Added https://github.com/ISISComputingGroup/IBEX/issues/2781 for reconnecting, I think it's a better solution to do it via some kind of heartbeat, which is a bit more work than the scope of this ticket.
* I've moved the strings to static variables, I agree that enums would be nicer but it's hard to implement as errors could also consist of underlying ZMQ issues or error messages sent back from the server. If you can think of a better way to do it please let me know, I spent a long time trying to find a tidier solution.
* Added the error logging back in
* Fixed inline comments

Thank you!

ThomasLohnert commented:
Finished reviewing, but a merge will require some follow up tasks such as updating our Nicos repo (not trivial). I will have a look at how to do this together with @DominicOram to learn more about the script server when I am back from leave after next week, so will hold off merging this until then.
